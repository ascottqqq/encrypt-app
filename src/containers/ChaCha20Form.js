import React, {Component} from 'react'
import FormErrors from '../components/FormErrors'
import 'bootstrap/dist/css/bootstrap.css'
import 'bootstrap/dist/css/bootstrap-theme.css'

class ChaCha20Form extends Component {
  constructor(props) {
    super(props)
    this.state = {key: '', keyValid: false, counter: '', nonce: '', 
      nonceValid: false, text: '', result: '', 
      formErrors: {hasErrors: false, key: '', nonce: '', serverErrors: ''}, 
      formValid: false, isEncrypt: false}
      this.processData = this.processData.bind(this)
      this.processError = this.processError.bind(this)
      this.handleSubmit = this.handleSubmit.bind(this)
  }

  handleOnKeyDown = (e) => {
    if (e.key === '-' || e.key === 'e' || e.key === '.' || e.key === '+') {
      e.preventDefault();
    }
  }

  handleTextInput = (e) => {
    const name = e.target.name
    const value = e.target.value
    this.setState({[name]: value},
      () => {this.validateField(name, value)})	
  }

  isBase64 = (str) => {
    try {
      return window.btoa(window.atob(str)) === str
    } catch(err) {
      return false
    }
  }
  
  processData(base64Result) {
    let formErrors = this.state.formErrors
    if (this.state.isEncrypt) {
      debugger
      formErrors.serverErrors = ''
      formErrors.hasErrors = formErrors.key.length > 0 || 
        formErrors.nonce.length > 0
      this.setState ({result: base64Result, formErrors: formErrors})
    } else {
      debugger
      if (base64Result) {
        let ascii = new Buffer(base64Result, 'base64').toString('ascii')
        this.setState ({result: ascii})
      } else {
        console.log("result is undefined")
      }
    }
  }

  processError(error) {
    let formErrors = this.state.formErrors
    formErrors.serverErrors = ' ' + error.message
    formErrors.hasErrors = true
    this.setState ({formErrors: formErrors})
  } 

  validateField(fieldName, value) {
    let fieldValidationErrors = this.state.formErrors
    let keyValid = this.state.keyValid
    let nonceValid = this.state.nonceValid
    let formValid = this.state.formValid
    switch(fieldName) {
      case 'key':
        const keyMax = "1157920892373161954235709850086879078532699846656" +
          "40564039457584007913129639936"
        let keyPad = "0".repeat(Math.max(keyMax.length, value.length))
        let paddedKeyValue = keyPad.substring(0, keyPad.length - 
          value.length) + value
        let paddedMaxKey = keyPad.substring(0, keyPad.length - 
          keyMax.length) + keyMax
        keyValid = !isNaN(value) && value !== "" &&
           paddedKeyValue < paddedMaxKey
        fieldValidationErrors.key = keyValid ? '' : ' is invalid'
        break
      case 'nonce':
        const nonceMax = "79228162514264337593543950336"
        let noncePad = "0".repeat(Math.max(nonceMax.length, value.length))
        let paddedNonceValue = noncePad.substring(0, noncePad.length - 
          value.length) + value
        let paddedMaxNonce = noncePad.substring(0, noncePad.length - 
          nonceMax.length) + nonceMax
        nonceValid = !isNaN(value) && value !== "" &&
                  paddedNonceValue < paddedMaxNonce
        fieldValidationErrors.nonce = nonceValid ? '' : ' is invalid'
        break
      default:
        break
    }
    fieldValidationErrors.hasErrors = 
      fieldValidationErrors.key.length > 0 || 
      fieldValidationErrors.nonce.length > 0 ||
      fieldValidationErrors.serverErrors.length > 0
    formValid = keyValid && nonceValid && this.state.text.length > 0
    this.setState({formErrors: fieldValidationErrors,
         keyValid: keyValid, nonceValid: nonceValid, 
         formValid: formValid})
  }

  handleSubmit(e) {
    debugger
    this.props.handleSubmit.call(this, e)
  }

  render() {
    var cardStyle = {width: "20rem"}
    return (
      <div className={this.state.formErrors.hasErrors ? 
          "panel panel-danger" : 
          "panel panel-default"}>
        <div className="panel-heading">
          <FormErrors formErrors={this.state.formErrors} />
        </div>
        <div className="panel-body">
        <div className="form-group">
          <label htmlFor="key-input">Key:</label>
          <input type="number"
            className="form-control" value={this.state.key} 
            name="key" id="key-input" onChange={this.handleTextInput} 
            onKeyDown={this.handleOnKeyDown} min="0" />
        </div>
        <div className="form-group">
          <label htmlFor="nonce-input">Nonce:</label>
          <input type="number" className="form-control" value={this.state.nonce} 
            name="nonce" id="nonce-input" onChange={this.handleTextInput} 
            onKeyDown={this.handleOnKeyDown} min="0" />
        </div>
        <div className="form-group">
          <label htmlFor="text-input">Text:</label>
          <textarea value={this.state.plaintext} className="form-control" 
            name="text" id="text-input" onChange={this.handleTextInput} />
        </div>
        <div className="btn-group" role="group">
          <button onClick={this.handleSubmit} className="btn btn-primary" name="encrypt"
            disabled={!this.state.formValid}>
            Encrypt
          </button>
          <button onClick={this.handleSubmit} className="btn btn-primary" name="decrypt"
            disabled={!this.state.formValid}>
            Decrypt
          </button>
        </div>
        <div className="card">
          <div className="card-body" style={cardStyle}>
            <h4 className="card-title">Result:</h4>
            <p className="card-text">{this.state.result}</p>
          </div>
        </div>
        </div>
      </div>
    )
  }
} export default ChaCha20Form 