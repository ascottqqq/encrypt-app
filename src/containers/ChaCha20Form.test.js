import React from 'react'
import {mount} from 'enzyme'
import ChaCha20Form from './ChaCha20Form'
import ReactDOM from 'react-dom'
import ReactTestUtils from 'react-dom/test-utils'
import renderer from 'react-test-renderer'

it('should handle correct input in key field', () => {
   let node = mount(<ChaCha20Form />);
   node.setState({key: '', keyValid: false, counter: '', nonce: '', 
      nonceValid: false, text: '', result: '', 
      formErrors: {hasErrors: false, key: '', nonce: '', serverErrors: ''}, 
      formValid: false, isEncrypt: false})

   let key = node.find('[name="key"]');
   let value = {target: {value: '10', name: 'key'}}

   key.value = '10'
   key.simulate('change', value )
   expect(node.state().key).toEqual('10')
   expect(node.state().formErrors.hasErrors).toEqual(false)
   expect(node.state().formErrors.key).toEqual('')
   expect(node.state().formValid).toEqual(false)
});

it('should handle correct input in nonce field', () => {
   let node = mount(<ChaCha20Form />);
   node.setState({key: '', keyValid: false, counter: '', nonce: '', 
      nonceValid: false, text: '', result: '', 
      formErrors: {hasErrors: false, key: '', nonce: '', serverErrors: ''}, 
      formValid: false, isEncrypt: false})

   let nonce = node.find('[name="nonce"]');
   let value = {target: {value: '11', name: 'nonce'}}

   nonce.value = '11'
   nonce.simulate('change', value )
   expect(node.state().nonce).toEqual('11')
   expect(node.state().formErrors.hasErrors).toEqual(false)
   expect(node.state().formErrors.nonce).toEqual('')
   expect(node.state().formValid).toEqual(false)
});

it('should handle correct input in text field', () => {
   let node = mount(<ChaCha20Form />);
   node.setState({key: '', keyValid: false, counter: '', nonce: '', 
      nonceValid: false, text: '', result: '', 
      formErrors: {hasErrors: false, key: '', nonce: '', serverErrors: ''}, 
      formValid: false, isEncrypt: false})

   let text = node.find('[name="text"]');
   let value = {target: {value: 'Hello', name: 'text'}}

   text.value = 'Hello'
   text.simulate('change', value )
   expect(node.state().text).toEqual('Hello')
   expect(node.state().formErrors.hasErrors).toEqual(false)
   expect(node.state().formValid).toEqual(false)
});

it('should handle correct input in text field', () => {
   let node = mount(<ChaCha20Form />);
   node.setState({key: '', keyValid: false, counter: '', nonce: '', 
      nonceValid: false, text: '', result: '', 
      formErrors: {hasErrors: false, key: '', nonce: '', serverErrors: ''}, 
      formValid: false, isEncrypt: false})

   let text = node.find('[name="text"]');
   let value = {target: {value: 'Hello', name: 'text'}}

   text.value = 'Hello'
   text.simulate('change', value )
   expect(node.state().text).toEqual('Hello')
   expect(node.state().formErrors.hasErrors).toEqual(false)
   expect(node.state().formValid).toEqual(false)
});

it('should handle correct input in key, nonce, and text fields', () => {
   let handleSubmitMock = jest.fn()
   let node = mount(<ChaCha20Form />);
   node.setState({key: '', keyValid: false, counter: '', nonce: '', 
      nonceValid: false, text: '', result: '', 
      formErrors: {hasErrors: false, key: '', nonce: '', serverErrors: ''}, 
      formValid: false, isEncrypt: false})

   let field = node.find('[name="key"]');
   let value = {target: {value: '1234', name: 'key'}}

   field.value = '1234'
   field.simulate('change', value )

   field = node.find('[name="nonce"]');
   value = {target: {value: '4321', name: 'nonce'}}

   field.value = '4321'
   field.simulate('change', value )

   field = node.find('[name="text"]');
   value = {target: {value: 'Hello World!', name: 'text'}}

   field.value = 'Hello World!'
   field.simulate('change', value )

   expect(handleSubmitMock.mock.calls.length).toBe(0)
   expect(node.state().key).toEqual('1234')
   expect(node.state().nonce).toEqual('4321')
   expect(node.state().text).toEqual('Hello World!')
   expect(node.state().formErrors.hasErrors).toEqual(false)
   expect(node.state().formValid).toEqual(true)
});

it('should handle encrypt button click correctly', () => {
   let handleSubmitMock = jest.fn()
   let node = mount(<ChaCha20Form handleSubmit={handleSubmitMock} />);
   node.setState({key: '', keyValid: false, counter: '', nonce: '', 
      nonceValid: false, text: '', result: '', 
      formErrors: {hasErrors: false, key: '', nonce: '', serverErrors: ''}, 
      formValid: false, isEncrypt: false})

   let field = node.find('[name="key"]');
   let value = {target: {value: '1234', name: 'key'}}

   field.value = '1234'
   field.simulate('change', value )

   field = node.find('[name="nonce"]');
   value = {target: {value: '4321', name: 'nonce'}}

   field.value = '4321'
   field.simulate('change', value )

   field = node.find('[name="text"]');
   value = {target: {value: 'Hello World!', name: 'text'}}

   field.value = 'Hello World!'
   field.simulate('change', value )

   field = node.find('[name="encrypt"]');
   value = {target: { name: 'encrypt'}}
   field.simulate('click', value)

   expect(handleSubmitMock.mock.calls.length).toBe(1)
   expect(handleSubmitMock.mock.calls[0][0].target.name).toEqual('encrypt')
   expect(node.state().key).toEqual('1234')
   expect(node.state().nonce).toEqual('4321')
   expect(node.state().text).toEqual('Hello World!')
   expect(node.state().formErrors.hasErrors).toEqual(false)
   expect(node.state().formValid).toEqual(true)
});

it('Should handle decrypt button click correctly', () => {
   let handleSubmitMock = jest.fn()
   let node = mount(<ChaCha20Form handleSubmit={handleSubmitMock} />);
   node.setState({key: '', keyValid: false, counter: '', nonce: '', 
      nonceValid: false, text: '', result: '', 
      formErrors: {hasErrors: false, key: '', nonce: '', serverErrors: ''}, 
      formValid: false, isEncrypt: false})

   let field = node.find('[name="key"]');
   let value = {target: {value: '1234', name: 'key'}}

   field.value = '1234'
   field.simulate('change', value )

   field = node.find('[name="nonce"]');
   value = {target: {value: '4321', name: 'nonce'}}

   field.value = '4321'
   field.simulate('change', value )

   field = node.find('[name="text"]');
   value = {target: {value: 'Hello World!', name: 'text'}}

   field.value = 'Hello World!'
   field.simulate('change', value )

   field = node.find('[name="decrypt"]');
   value = {target: { name: 'decrypt'}}
   field.simulate('click', value)

   expect(handleSubmitMock.mock.calls.length).toBe(1)
   expect(handleSubmitMock.mock.calls[0][0].target.name).toEqual('decrypt')
   expect(node.state().key).toEqual('1234')
   expect(node.state().nonce).toEqual('4321')
   expect(node.state().text).toEqual('Hello World!')
   expect(node.state().formErrors.hasErrors).toEqual(false)
   expect(node.state().formValid).toEqual(true)
});

it('should handle error input in key field', () => {
   let node = mount(<ChaCha20Form />);
   node.setState({key: '', keyValid: false, counter: '', nonce: '', 
      nonceValid: false, text: '', result: '', 
      formErrors: {hasErrors: false, key: '', nonce: '', serverErrors: ''}, 
      formValid: false, isEncrypt: false})

   let key = node.find('[name="key"]');
   let value = {target: {value: '115792089237316195423570985008687907853269' +
     '984665640564039457584007913129639936', name: 'key'}}

   key.simulate('change', value )
   expect(node.state().formErrors.key).toEqual(' is invalid')
   expect(node.state().formErrors.hasErrors).toEqual(true)
   expect(node.state().formValid).toEqual(false)
});

it('Should handle error input in nonce field', () => {
   let node = mount(<ChaCha20Form />);
   node.setState({key: '', keyValid: false, counter: '', nonce: '', 
      nonceValid: false, text: '', result: '', 
      formErrors: {hasErrors: false, key: '', nonce: '', serverErrors: ''}, 
      formValid: false, isEncrypt: false})

   let nonce = node.find('[name="nonce"]');
   let value = {target: {value: '79228162514264337593543950336', 
     name: 'nonce'}}

   nonce.simulate('change', value )
   expect(node.state().formErrors.nonce).toEqual(' is invalid')
   expect(node.state().formErrors.hasErrors).toEqual(true)
   expect(node.state().formValid).toEqual(false)
});

it('renders without error correctly', () => {
  const tree = renderer.create(<ChaCha20Form />).toJSON()
  expect(tree).toMatchSnapshot()
});

it('renders with valid data correctly', () => {
  const container = renderer.create(<ChaCha20Form />)
  container.getInstance().handleTextInput({target: { name: 'key', 
    value: '55'}})
  container.getInstance().handleTextInput({target: { name: 'nonce', 
    value: '66'}})
  container.getInstance().handleTextInput({target: { name: 'text', 
    value: 'Ready'}})
  expect(container.toJSON()).toMatchSnapshot()
});

it('renders with error correctly', () => {
  const container = renderer.create(<ChaCha20Form />)
  container.getInstance().validateField('nonce', '79228162514264337' + 
    '593543950336')
  expect(container.toJSON()).toMatchSnapshot();
});

it('renders with server error correctly', () => {
  const container = renderer.create(<ChaCha20Form />)
  container.getInstance().processError({message: 'Server down'})
  expect(container.toJSON()).toMatchSnapshot();
});