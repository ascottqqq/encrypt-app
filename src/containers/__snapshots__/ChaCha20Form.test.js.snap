// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with error correctly 1`] = `
<div
  className="panel panel-danger"
>
  <div
    className="panel-heading"
  >
    <div
      className=""
    >
      
      
      <p>
        Nonce
         is invalid
      </p>
      
    </div>
  </div>
  <div
    className="panel-body"
  >
    <div
      className="form-group"
    >
      <label
        htmlFor="key-input"
      >
        Key:
      </label>
      <input
        className="form-control"
        id="key-input"
        min="0"
        name="key"
        onChange={[Function]}
        onKeyDown={[Function]}
        type="number"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="nonce-input"
      >
        Nonce:
      </label>
      <input
        className="form-control"
        id="nonce-input"
        min="0"
        name="nonce"
        onChange={[Function]}
        onKeyDown={[Function]}
        type="number"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="text-input"
      >
        Text:
      </label>
      <textarea
        className="form-control"
        id="text-input"
        name="text"
        onChange={[Function]}
        value={undefined}
      />
    </div>
    <div
      className="btn-group"
      role="group"
    >
      <button
        className="btn btn-primary"
        disabled={true}
        name="encrypt"
        onClick={[Function]}
      >
        Encrypt
      </button>
      <button
        className="btn btn-primary"
        disabled={true}
        name="decrypt"
        onClick={[Function]}
      >
        Decrypt
      </button>
    </div>
    <div
      className="card"
    >
      <div
        className="card-body"
        style={
          Object {
            "width": "20rem",
          }
        }
      >
        <h4
          className="card-title"
        >
          Result:
        </h4>
        <p
          className="card-text"
        >
          
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders with server error correctly 1`] = `
<div
  className="panel panel-danger"
>
  <div
    className="panel-heading"
  >
    <div
      className=""
    >
      
      
      
      <p>
        ServerErrors
         Server down
      </p>
    </div>
  </div>
  <div
    className="panel-body"
  >
    <div
      className="form-group"
    >
      <label
        htmlFor="key-input"
      >
        Key:
      </label>
      <input
        className="form-control"
        id="key-input"
        min="0"
        name="key"
        onChange={[Function]}
        onKeyDown={[Function]}
        type="number"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="nonce-input"
      >
        Nonce:
      </label>
      <input
        className="form-control"
        id="nonce-input"
        min="0"
        name="nonce"
        onChange={[Function]}
        onKeyDown={[Function]}
        type="number"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="text-input"
      >
        Text:
      </label>
      <textarea
        className="form-control"
        id="text-input"
        name="text"
        onChange={[Function]}
        value={undefined}
      />
    </div>
    <div
      className="btn-group"
      role="group"
    >
      <button
        className="btn btn-primary"
        disabled={true}
        name="encrypt"
        onClick={[Function]}
      >
        Encrypt
      </button>
      <button
        className="btn btn-primary"
        disabled={true}
        name="decrypt"
        onClick={[Function]}
      >
        Decrypt
      </button>
    </div>
    <div
      className="card"
    >
      <div
        className="card-body"
        style={
          Object {
            "width": "20rem",
          }
        }
      >
        <h4
          className="card-title"
        >
          Result:
        </h4>
        <p
          className="card-text"
        >
          
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders with valid data correctly 1`] = `
<div
  className="panel panel-default"
>
  <div
    className="panel-heading"
  >
    <div
      className="text-center"
    >
      <p>
        ChaCha 20
      </p>
    </div>
  </div>
  <div
    className="panel-body"
  >
    <div
      className="form-group"
    >
      <label
        htmlFor="key-input"
      >
        Key:
      </label>
      <input
        className="form-control"
        id="key-input"
        min="0"
        name="key"
        onChange={[Function]}
        onKeyDown={[Function]}
        type="number"
        value="55"
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="nonce-input"
      >
        Nonce:
      </label>
      <input
        className="form-control"
        id="nonce-input"
        min="0"
        name="nonce"
        onChange={[Function]}
        onKeyDown={[Function]}
        type="number"
        value="66"
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="text-input"
      >
        Text:
      </label>
      <textarea
        className="form-control"
        id="text-input"
        name="text"
        onChange={[Function]}
        value={undefined}
      />
    </div>
    <div
      className="btn-group"
      role="group"
    >
      <button
        className="btn btn-primary"
        disabled={false}
        name="encrypt"
        onClick={[Function]}
      >
        Encrypt
      </button>
      <button
        className="btn btn-primary"
        disabled={false}
        name="decrypt"
        onClick={[Function]}
      >
        Decrypt
      </button>
    </div>
    <div
      className="card"
    >
      <div
        className="card-body"
        style={
          Object {
            "width": "20rem",
          }
        }
      >
        <h4
          className="card-title"
        >
          Result:
        </h4>
        <p
          className="card-text"
        >
          
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders without error correctly 1`] = `
<div
  className="panel panel-default"
>
  <div
    className="panel-heading"
  >
    <div
      className="text-center"
    >
      <p>
        ChaCha 20
      </p>
    </div>
  </div>
  <div
    className="panel-body"
  >
    <div
      className="form-group"
    >
      <label
        htmlFor="key-input"
      >
        Key:
      </label>
      <input
        className="form-control"
        id="key-input"
        min="0"
        name="key"
        onChange={[Function]}
        onKeyDown={[Function]}
        type="number"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="nonce-input"
      >
        Nonce:
      </label>
      <input
        className="form-control"
        id="nonce-input"
        min="0"
        name="nonce"
        onChange={[Function]}
        onKeyDown={[Function]}
        type="number"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="text-input"
      >
        Text:
      </label>
      <textarea
        className="form-control"
        id="text-input"
        name="text"
        onChange={[Function]}
        value={undefined}
      />
    </div>
    <div
      className="btn-group"
      role="group"
    >
      <button
        className="btn btn-primary"
        disabled={true}
        name="encrypt"
        onClick={[Function]}
      >
        Encrypt
      </button>
      <button
        className="btn btn-primary"
        disabled={true}
        name="decrypt"
        onClick={[Function]}
      >
        Decrypt
      </button>
    </div>
    <div
      className="card"
    >
      <div
        className="card-body"
        style={
          Object {
            "width": "20rem",
          }
        }
      >
        <h4
          className="card-title"
        >
          Result:
        </h4>
        <p
          className="card-text"
        >
          
        </p>
      </div>
    </div>
  </div>
</div>
`;
